arroyo-worker/src/operators/async_map.rs

```diff
@@ -183,13 +199,69 @@ impl<
         }
     }
 
+    async fn handle_watermark(
+        &mut self,
+        watermark: arroyo_types::Watermark,
+        _ctx: &mut crate::engine::Context<InKey, OutT>,
+    ) {
+        self.watermarks.push_back((self.next_id, watermark));
+    }
+
     fn tables(&self) -> Vec<TableDescriptor> {
         vec![arroyo_state::global_table("a", "AsyncMapOperator state")]
     }
 
-    async fn on_close(&mut self, _ctx: &mut Context<InKey, OutT>) {
+    async fn on_close(
+        &mut self,
+        ctx: &mut Context<InKey, OutT>,
+        final_message: &Option<Message<InKey, OutT>>,
+    ) {
+        if let Some(Message::EndOfData) = final_message {
+            debug!(
+                "AsyncMapOperator end of data with {} futures",
+                self.futures.len()
+            );
+            while let Some((id, result)) = self.futures.next().await {
+                self.handle_future(id, result, ctx).await;
+            }
+        }
         self.udf_context.close().await;
     }
+
+    async fn handle_future(
+        &mut self,
+        id: usize,
+        result: Result<OutT, Elapsed>,
+        ctx: &mut Context<InKey, OutT>,
+    ) {
+        match result {
+            Ok(value) => {
+                let index = self.inputs.len() - (self.next_id - id);
+                let input = self.inputs[index].clone().unwrap();
+                ctx.collector
+                    .collect(Record {
+                        timestamp: input.timestamp,
+                        key: input.key.clone(),
+                        value,
+                    })
+                    .await;
+                self.on_collect(id, ctx).await;
+            }
+            Err(_e) => {
+                if self.futures.is_ordered() {
+                    unimplemented!(
+                        "Ordered Async UDF timed out, currently panic to preserve ordering"
+                    );
+                }
+                warn!("Unordered Async UDF timed out, retrying");
+                self.futures.push_back((self.udf)(
```

Comment:
Is there anything to prevent an infinite retry loop?

Reply:
There isn't currently, we should build out the options here in the next version.